source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,198 @@
+#include "adaptive_load/adaptive_load_controller_impl.h"
+
+#include <chrono>
+
+#include "envoy/common/exception.h"
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "nighthawk/adaptive_load/adaptive_load_controller.h"
+#include "nighthawk/adaptive_load/metrics_plugin.h"
+#include "nighthawk/adaptive_load/scoring_function.h"
+#include "nighthawk/adaptive_load/step_controller.h"
+
+#include "external/envoy/source/common/common/logger.h"
+#include "external/envoy/source/common/common/statusor.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "api/adaptive_load/adaptive_load.pb.h"
+#include "api/adaptive_load/benchmark_result.pb.h"
+#include "api/adaptive_load/metric_spec.pb.h"
+#include "api/client/options.pb.h"
+#include "api/client/output.pb.h"
+#include "api/client/service.grpc.pb.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/status.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "adaptive_load/metrics_plugin_impl.h"
+#include "adaptive_load/plugin_loader.h"
+
+namespace Nighthawk {
+
+namespace {
+
+using nighthawk::adaptive_load::AdaptiveLoadSessionOutput;
+using nighthawk::adaptive_load::AdaptiveLoadSessionSpec;
+using nighthawk::adaptive_load::BenchmarkResult;
+using nighthawk::adaptive_load::MetricEvaluation;
+using nighthawk::adaptive_load::MetricSpec;
+using nighthawk::adaptive_load::MetricSpecWithThreshold;
+using nighthawk::adaptive_load::ThresholdSpec;
+
+/**
+ * Loads and initializes MetricsPlugins requested in the session spec. Assumes the spec has already
+ * been validated; crashes the process otherwise.
+ *
+ * @param spec Adaptive load session spec proto that has already been validated.
+ *
+ * @return Map from MetricsPlugin names to initialized plugins, to be used in the course of a single
+ * adaptive load session based on the session spec.
+ */
+absl::flat_hash_map<std::string, MetricsPluginPtr>
+LoadMetricsPlugins(const AdaptiveLoadSessionSpec& spec) {
+  absl::flat_hash_map<std::string, MetricsPluginPtr> name_to_custom_metrics_plugin_map;
+  for (const envoy::config::core::v3::TypedExtensionConfig& config :
+       spec.metrics_plugin_configs()) {
+    absl::StatusOr<MetricsPluginPtr> metrics_plugin_or = LoadMetricsPlugin(config);
```

Comment:
nit: usually it's better to perform the load right after validations and pass in the metric plugins to avoid the assert. No action needed if it's too much of code refactoring.

Reply:
Agreed, it's a bit awkward for historical reasons -- filed #547